### PR TITLE
Add types, make deserialization work, fix UTXOInput

### PR DESCRIPTION
**Pull request by Thoralf-M**
# Description of change

Add types, make deserialization work from block down to all types inside, fix UTXOInput in nodejs binding

## Links to any relevant issues

#42 

## Type of change

- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How the change has been tested

Added test and run examples

## Change checklist

Tick the boxes that are relevant to your changes, and delete any items that are not.

- [ ] I have followed the contribution guidelines for this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that new and existing unit tests pass locally with my changes

### CODE REVIEW

File: bindings/python/iota_sdk/client/client.py [hunk 1]
```diff
@@ -502,27 +506,29 @@ def unhealthy_nodes(self) -> List[Dict[str, Any]]:
         """
         return self._call_method('unhealthyNodes')
 
-    def prepare_transaction(self, secret_manager=None, options=None):
+    def prepare_transaction(self, secret_manager: Optional[LedgerNanoSecretManager | MnemonicSecretManager |
+                            SeedSecretManager | StrongholdSecretManager] = None, options=None) -> PreparedTransactionData:
         """Prepare a transaction for signing.
         """
-        return self._call_method('prepareTransaction', {
+        return from_dict(PreparedTransactionData, self._call_method('prepareTransaction', {
```
Comment (DaughterOfMars) [inline bindings/python/iota_sdk/client/client.py:483]: Aren't these supposed to be without `Data` now?

Comment (Thoralf-M) [reply bindings/python/iota_sdk/client/client.py:483]: Only in the wallet in the account methods

**Review by DaughterOfMars — DISMISSED** [review]

File: bindings/python/iota_sdk/types/block.py [hunk 2]
```diff
@@ -17,20 +18,24 @@ class Block:
...
     payload: Optional[Payload] = None
 
     @classmethod
     def from_dict(cls, block_dict: Dict) -> Block:
-        obj = cls.__new__(cls)
-        super(Block, obj).__init__()
-        for k, v in block_dict.items():
-            setattr(obj, k, v)
-        return obj
+        return from_dict(Block, block_dict)
```
Comment (thibault-martinez) [inline bindings/python/iota_sdk/types/block.py:33]: I don't know Python but that kinda feels a bit recursive

Comment (thibault-martinez) [reply bindings/python/iota_sdk/types/block.py:33]: It's probably not but just checking

File: bindings/python/iota_sdk/types/input.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    """Represents an input referencing an output.
+
+    Attributes:
+        type: The type of input.
+        transactionId: The transaction id that created the output.
+        transactionOutputIndex: The output index.
+    """
+    type: int
+    transactionId: HexStr
+    transactionOutputIndex: int
```
Comment (thibault-martinez) [inline bindings/python/iota_sdk/types/input.py:32]: Is that really the serialized name?

Comment (Thoralf-M) [reply bindings/python/iota_sdk/types/input.py:32]: Yes https://api.testnet.shimmer.network/api/core/v2/blocks/0xc27f1bd5dbd719ca19d98a351eca6ff2e32bbd6791351f2475cc9fdff02d8323

File: bindings/python/iota_sdk/types/block.py (hunk 2 above)
Comment (Thoralf-M) [reply bindings/python/iota_sdk/types/block.py:33]: It's not :+1: Would be if it was `Block.from_dict()` but this way it's using the dacite import

File: bindings/python/iota_sdk/types/input.py (hunk 3 above)
Comment (thibault-martinez) [reply bindings/python/iota_sdk/types/input.py:32]: Weird, ok

File: bindings/python/iota_sdk/types/block.py (hunk 2 above)
Comment (thibault-martinez) [reply bindings/python/iota_sdk/types/block.py:33]: Thanks

File: bindings/python/iota_sdk/types/unlock.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+        return config
+
+
+@dataclass
+class SignatureUnlock:
+    """An unlock holding a signature unlocking one or more inputs.
+    """
+    signature: Ed25519Signature
+    transactionInputIndex: int
```
Comment (thibault-martinez) [inline bindings/python/iota_sdk/types/unlock.py]: What's this?

Comment (Thoralf-M) [reply bindings/python/iota_sdk/types/unlock.py]: Scam

Comment (Thoralf-M) [reply bindings/python/iota_sdk/types/unlock.py]: Removed

Comment (Thoralf-M) [reply bindings/python/iota_sdk/types/unlock.py]: I even built this type with this int to see if the serialization works :carlos:

File: bindings/python/iota_sdk/types/unlock.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    Nft = 3
+
+
+@dataclass
+class Unlock:
+    """Unlock type.
+    """
+    type: int
+    signature: Optional[Ed25519Signature] = None
+    transactionInputIndex: Optional[int] = None
```
Comment (thibault-martinez) [inline bindings/python/iota_sdk/types/unlock.py]: And what about this ?

Comment (Thoralf-M) [reply bindings/python/iota_sdk/types/unlock.py]: Also wrong, thanks

File: bindings/python/iota_sdk/types/unlock.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+            config['signature'] = config['signature'].__dict__
+
+        return config
+
+
+@dataclass
+class SignatureUnlock:
+    """An unlock holding a signature unlocking one or more inputs.
+    """
+    signature: Ed25519Signature
```
Comment (thibault-martinez) [inline bindings/python/iota_sdk/types/unlock.py:37]: Should technically be a `Signature`, maybe we don't have that yet. But if `Ed25519Signature` serialises it's type it's fine.

File: bindings/python/iota_sdk/types/transaction_data.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+@dataclass
+class SignedTransactionData:
+    """Helper class for offline signing.
+
+    Attributes:
+        transactionPayload: The transaction payload.
+        inputsData: Data about the inputs consumed in the transaction.
+    """
+    # TODO: have TransactionPayload here?
```
Comment (thibault-martinez) [inline bindings/python/iota_sdk/types/transaction_data.py]: I assume it does make a difference so we need to be sure

File: bindings/python/iota_sdk/types/transaction_data.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+@dataclass
+class PreparedTransactionData:
+    """Helper class for offline signing.
+
+    Attributes:
+        essence: The transaction essence.
+        inputsData: Data about the inputs which is required for signing.
+        remainder: Data about a remainder.
+    """
+    essence: RegularTransactionEssence
```
Comment (thibault-martinez) [inline bindings/python/iota_sdk/types/transaction_data.py:50]: Should be a `TransactionEssence` but probably fine for now

File: bindings/python/iota_sdk/types/transaction_data.py (hunk 7 above)
Comment (Thoralf-M) [reply bindings/python/iota_sdk/types/transaction_data.py]: Updated, commented that when I removed the TransactionPayload completely :trollface:

File: bindings/python/iota_sdk/types/transaction_data.py (hunk 8 above)
Comment (Thoralf-M) [reply bindings/python/iota_sdk/types/transaction_data.py:50]: Can add this tomorrow

File: bindings/python/iota_sdk/types/unlock.py (hunk 6 above)
Comment (thibault-martinez) [reply bindings/python/iota_sdk/types/unlock.py:37]: Will do later?

File: bindings/python/iota_sdk/types/payload.py [hunk 9]
```diff
@@ -3,50 +3,88 @@
...
+    """Initialize a Payload.
+    """
+    type: int
+    # tagged data
+    tag: Optional[HexStr] = None
+    data: Optional[HexStr] = None
+    # transaction
+    essence: Optional[RegularTransactionEssence] = None
+    unlocks: Optional[List[Unlock]] = None
+    # milestone
```
Comment (thibault-martinez) [inline bindings/python/iota_sdk/types/payload.py]: But milestone also had an essence field? How does that work?

Comment (Thoralf-M) [reply bindings/python/iota_sdk/types/payload.py]: There is no essence field in the dto

File: bindings/python/iota_sdk/types/unlock.py (hunk 6 above)
Comment (Thoralf-M) [reply bindings/python/iota_sdk/types/unlock.py:37]: Added it to the issue

File: bindings/python/iota_sdk/types/payload.py (hunk 9 above)
Comment (thibault-martinez) [reply bindings/python/iota_sdk/types/payload.py]: Oh ok

File: bindings/python/iota_sdk/types/payload.py [hunk 10]
```diff
@@ -3,50 +3,80 @@
...
+            lambda x: x.as_dict(), config['outputs']))
+
+        return config
+
+
+@dataclass
+class Payload():
+    """Initialize a Payload.
+    """
+    type: int
```
Comment (thibault-martinez) [inline bindings/python/iota_sdk/types/payload.py:64]: So much better :)

Comment (Thoralf-M) [reply bindings/python/iota_sdk/types/payload.py:64]: Yeah found the nice `field()` thing from dataclasses which makes this possible
Will update the remaining classes in another PR

**Review by thibault-martinez — APPROVED** [review]

**Review by thibault-martinez — CHANGES_REQUESTED** [review]
Conflicts

**Review by thibault-martinez — APPROVED** [review]

**Review by kwek20 — APPROVED** [review]
Not sure if tests/test_block.py is really needed, seems more of a tool you used to test if it works

**Pull request merged**